pkg/bgpv1/manager/reconcilerv2/neighbor.go

```diff
@@ -0,0 +1,362 @@
+			break
+		}
+	}
+
+	neighMetadata[instanceName] = peers
+}
+
+func (r *NeighborReconciler) Name() string {
+	return "Neighbor"
+}
+
+// Priority of neighbor reconciler is higher than pod/service announcements.
+// This is important for graceful restart case, where all expected routes are pushed
+// into gobgp RIB before neighbors are added. So, gobgp can send out all prefixes
+// within initial update message exchange with neighbors before sending EOR marker.
+func (r *NeighborReconciler) Priority() int {
+	return 60
+}
+
+func (r *NeighborReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
+	if p.DesiredConfig == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil CiliumBGPNodeInstance")
+	}
+	if p.BGPInstance == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil BGPInstance")
+	}
+
+	var (
+		l = log.WithFields(
+			logrus.Fields{
+				"component": fmt.Sprintf("reconciler.v2.%s", r.Name()),
+			},
+		)
+
+		toCreate []*PeerData
+		toRemove []*PeerData
+		toUpdate []*PeerData
+		curNeigh []*PeerData = nil
+	)
+	newNeigh := p.DesiredConfig.Peers
+	l.Debugf("Begin reconciling peers for virtual router %s", p.DesiredConfig.Name)
+
+	// get current configured peers
+	curInstance := r.getMetadata(p.BGPInstance)
+	if curInstance != nil {
+		curNeigh = curInstance[p.DesiredConfig.Name]
+	}
+
+	type member struct {
+		new *PeerData
+		cur *PeerData
+	}
+
+	nset := map[string]*member{}
+
+	for i, n := range newNeigh {
+		var (
+			key = n.Name
+			h   *member
+			ok  bool
+		)
+
+		config, err := r.getPeerConfig(n.PeerConfigRef.Name)
+		if err != nil {
+			return err
+		}
+
+		passwd, err := r.getPeerPassword(config)
+		if err != nil {
+			return err
+		}
+
+		if h, ok = nset[key]; !ok {
+			nset[key] = &member{
+				new: &PeerData{
+					Peer:     &newNeigh[i],
+					Config:   config,
+					Password: passwd,
+				},
+			}
+			continue
+		}
+		h.new = &PeerData{
+			Peer:     &newNeigh[i],
+			Config:   config,
+			Password: passwd,
+		}
+	}
+
+	for i, n := range curNeigh {
+		var (
+			key = n.Peer.Name
+			h   *member
+			ok  bool
+		)
+
+		if h, ok = nset[key]; !ok {
+			nset[key] = &member{
+				cur: curNeigh[i],
+			}
+			continue
+		}
+		h.cur = curNeigh[i]
+	}
+
+	for _, m := range nset {
+		// present in new neighbors (set new) but not in current neighbors (set cur)
+		if m.new != nil && m.cur == nil {
+			toCreate = append(toCreate, m.new)
+		}
+		// present in current neighbors (set cur) but not in new neighbors (set new)
+		if m.cur != nil && m.new == nil {
+			toRemove = append(toRemove, m.cur)
+		}
+		// present in both new neighbors (set new) and current neighbors (set cur), update if they are not equal
+		if m.cur != nil && m.new != nil {
+			if !m.cur.Equal(m.new) {
+				toUpdate = append(toUpdate, m.new)
+			}
+		}
+	}
+
+	if len(toCreate) > 0 || len(toRemove) > 0 || len(toUpdate) > 0 {
+		l.Infof("Reconciling peers for virtual router with %s", p.DesiredConfig.Name)
```

Comment:
I am fixing all logs to have k/v pairs instead of string format. 
Yes, having metrics hooks into reconcilers will be useful. Each reconciler can expose appropriate statistics.